### ISSUE TITLE
Pubsub mechanism for Parachains

### ISSUE THREAD

**Issue by bkchr** [paritytech/polkadot-sdk#606]
When parachains need to exchange information, they currently rely on sending XCM messages to each other or resorting to an off-chain protocol to obtain a storage proof from another chain. However, these solutions are complex and expensive, particularly when data should be disseminated amongst a large group of parachains, such as the price of DOT. 

A potential solution would be the introduction of a publish-subscribe (pub-sub) mechanism. This would enable parachains to subscribe to `(para_id, key)` information within their runtime. Collators would then be responsible for integrating this information into the runtime during block creation. The information would be published by parachains as an output of their validation process, necessitating changes to the `ValidationResult` as specified in [issue paritytech/polkadot#7191](https://github.com/paritytech/polkadot/issues/7191). 

The `(k, v)` information would be stored in the relay chain state, and it would likely be most efficient to maintain a separate child trie for each `para_id`. This would allow a collator to notify the parachain runtime when a change occurs in one of the subscribed items. 

However, the parachain runtime cannot fully trust the collator to include the data. Therefore, we only want to include the data when changes occur to conserve space. In such cases, the collator could include the child trie root (as part of the relay chain proof), which would be sufficient to prove that no changes occurred or provide proof of the child trie with the changed data.

CC @gavofyork

**Comment by vstam1** [comment]
Hi @bkchr, 
I was wondering if you have an update on the pubsub mechanism? I think a lot of parachains would really benefit from this feature, and I was wondering if there is a way we can contribute? 
Thanks!

**Comment by JuaniRios** [comment]
I also could really use this feature

**Comment by wischli** [comment]
Same here!

**Comment by bkchr** [comment]
Yeah we have talked a little bit "off github" about this. The current plan is the following:

1. Introduce a new XCM message to publish certain information. This will be send from the parachain to the relay chain.
2. Then have the parachain read the relay chain state and put it into the `ValidationData` inherent. (I'm working on this already, but on some kind of optimized version.)

If you need this, could you may look into 1? Providing 2 in some "dirty" way should be possible quite easily.

**Comment by vstam1** [comment]
Thanks for your quick response. I was wondering if there is another way of publishing the information in step 1 without XCM? Or at least without a new XCM instruction. This new instruction will then probably be introduced in XCM v5 and that can still take months. Might we be able to use the Transact instruction to publish the information?

**Comment by bkchr** [comment]
@franciscoaguirre what is the status on XCMv4? Still possible to sneak in some new instruction?

**Comment by franciscoaguirre** [comment]
We can. We wanted to get it out as soon as possible but the PR still doesn't have enough reviews. A new instruction wouldn't hurt

**Comment by bkchr** [comment]
Nice! Could we add some kind of instruction like this:
```
Publish { data: Vec<(Key, Value)>
```

While `Key` and `Value` just being `Vec<u8>`.

**Comment by franciscoaguirre** [comment]
We can work on the implementation here: https://github.com/paritytech/polkadot-sdk/pull/2359, it's pointing to the v4 branch.

In the meantime, I created an RFC for this to discuss how it should work: https://github.com/paritytech/xcm-format/pull/48
I feel like this could be a more generic instruction like `Store` that could potentially be used for storage parachains in the future. I see many people want this, please go and comment in there to make a nice new addition to the language :)

**Comment by franciscoaguirre** [comment]
Since XCMv4 was just a rename, this'll go in XCMv5. I urge you again to take a look at the RFC and comment. This will only go into v5 if we're all convinced about that RFC.

**Comment by metricaez** [comment]
About PersistedValidationData population

TL:DR - After extending PersistedValidationData with a new published_data field, block production stalls once I start populating it. The relay sees the field as non-empty while the parachain’s relay client still sees it as empty, leading to a mismatch. I’m trying to understand (1) where else PVD is being checked beyond `prospective-parachains` and `inclusion`, and (2) whether such a relay/parachain desync on `published_data` is something that has been evaluated or if I’m likely missing a critical population point.

Hey! We at Blockdeep are working on this system.

I’ve implemented a `broadcaster pallet` on the `relay` that stores `published_data` in tries and added a `Publish` XCM instruction to write into them. This part is working and already tested. I also extended `PersistedValidationData` to include a `published_data` field, which I populate inside `make_persisted_validation_data`. For `make_persisted_validation_data_with_parent` I haven’t added the ParaId parameter to avoid changing the signature, so instead I populate it right after inside `verify_backed_candidate`, which is the only place where that method or path is used. On the parachain side the plan is to eventually read the data back via new storage in `parachain-system`.

If I leave `published_data` empty on every `PersistedValidationData`, block production runs without any problems. The moment I actually populate it inside by publishing data through XCM, the parachain stalls. While debugging I noticed that at some point the relay client of the parachain lags behind and still “sees” an empty published_data, while the relay side shows it as non-empty. This mismatch seems to be the reason the chain stalls however I don't have any erro, parachain just stops proposing candidates.

My questions are: apart from the `prospective-parachains` and `inclusion` paths, where else are PersistedValidationData values being checked? I know parachain-system validates some parameters, but my understanding is that it shouldn’t break simply because published_data isn’t validated in the parachain runtime yet. And secondly, has the possibility of this desync been discussed before, a situation where from the parachain’s point of view the field is still empty while from the relay’s perspective it has already been populated? Or is it more likely that I am likely just forgetting or messing the population of the validation data at a critical point in the flow?

Any pointers into the right places to look would be appreciated, since PersistedValidationData ends up being threaded through many parts of block generation and as I do not see any particular log, it's being an interesting debugging process to say the least.

**Comment by bkchr** [comment]
You can not just modify `PersistedValidationData`. We don't support this and right now the best way to solve this is to just send the `XCM::Publish` to the relay chain as upwards message.

**Comment by metricaez** [comment]
Hey @bkchr, thanks for the follow up. I already had the data being published with Publish on the relay, my issue was on data propagation to parachains but thanks to your input I realized  I misunderstood the conversation, the idea of “…put it into the ValidationData inherent.” was not to put it inside `PersistedValidationData` but to use the inherent via `ParachainInherentData` to get the data to the parachain.

I took that path and now I am unblocked. 

Thanks

**Comment by bkchr** [comment]
> but to use the inherent via `ParachainInherentData` to get the data to the parachain.

Yeah, there you basically need to add more data to the relay chain proof, to load the data from there.

**Comment by metricaez** [comment]
> Yeah, there you basically need to add more data to the relay chain proof, to load the data from there.

I have a working PoC based on extending `ParachainInherentData` with a new field `published_data` as published data is stored on **child tries**, I populate it via runtime APIs when building the Inherent. Really similar to already existing `horizontal_messages`.

What I will send in the `relay_chain_state` is the root of each child trie which i am already storing on a `#[storage]` of my `broadcaster` pallet  and use as trusted source for validating that the data didn't change. Mainly because `relay_chain_state` get's values from the main relay storage, not child tries therefore for consistency I would like to keep `published_data` on another field of `ParachainInherentData`

@bkchr does this look okey to you ?

**Comment by bkchr** [comment]
The child trie proof can just be included in the main proof. No need to put this into an extra field.
